File: examples/python/02_proof_of_possession.py
"""Proof of possession: an audience-bound challenge/response round-trip.

The verifier issues a single-use challenge naming ITS OWN identifier as the
audience; the agent signs it; the verifier consumes the challenge (single
use) and checks the response against its own audience string — so a proof
collected by one service can never be replayed to another.

Run (after `pip install -e .` from the repo root):

    python examples/python/02_proof_of_possession.py
"""

from fg_agent_id import AgentIdentity, ChallengeStore

agent = AgentIdentity.generate("prover")

# Verifier side: issue a challenge bound to this verifier's identity.
store = ChallengeStore()
challenge = store.issue(audience="https://verifier.example")
print(f"challenge id  : {challenge.challenge_id}")

# Agent side: sign the challenge with the identity key.
response = challenge.respond(agent.keys, agent.address)

# Verifier side: consume (single use), then verify against OUR audience —
# never the audience the response claims.
issued = store.consume(response.challenge_id)
assert issued is not None
address = response.verify(issued, audience="https://verifier.example")
print(f"proved holder : {address}")
assert address == agent.address

# The challenge is gone now: a second use fails.
assert store.consume(response.challenge_id) is None
print("replay        : blocked (challenge is single-use)")

File: examples/js/02_proof_of_possession.mjs
/**
 * Proof of possession: an audience-bound challenge/response round-trip.
 *
 * Run from the repo root (build once with `cd js && npm install`):
 *
 *     node examples/js/02_proof_of_possession.mjs
 */

import { AgentIdentity, ChallengeStore } from "../../js/dist/src/index.js";
import assert from "node:assert/strict";

const agent = await AgentIdentity.generate("prover");

// Verifier side: issue a challenge bound to this verifier's identity.
const store = new ChallengeStore();
const challenge = store.issue("https://verifier.example");
console.log(`challenge id  : ${challenge.challengeId}`);

// Agent side: sign the challenge with the identity key.
const response = await challenge.respond(agent.keys, agent.address);

// Verifier side: consume (single use), then verify against OUR audience.
const issued = store.consume(response.challengeId);
assert.ok(issued);
const address = await response.verify(issued, "https://verifier.example");
console.log(`proved holder : ${address}`);
assert.equal(address, agent.address);

// The challenge is gone now: a second use fails.
assert.equal(store.consume(response.challengeId), null);
console.log("replay        : blocked (challenge is single-use)");
